storage/operation/pebbleimpl/writer.go

```diff
@@ -0,0 +1,83 @@
+package pebbleimpl
+
+import (
+	"github.com/cockroachdb/pebble"
+
+	"github.com/onflow/flow-go/storage"
+	op "github.com/onflow/flow-go/storage/operation"
+)
+
+type ReaderBatchWriter struct {
+	globalReader storage.Reader
+	batch        *pebble.Batch
+
+	callbacks op.Callbacks
+}
+
+var _ storage.ReaderBatchWriter = (*ReaderBatchWriter)(nil)
+
+func (b *ReaderBatchWriter) GlobalReader() storage.Reader {
+	return b.globalReader
+}
+
+func (b *ReaderBatchWriter) Writer() storage.Writer {
+	return b
+}
+
+func (b *ReaderBatchWriter) PebbleWriterBatch() *pebble.Batch {
+	return b.batch
+}
+
+func (b *ReaderBatchWriter) AddCallback(callback func(error)) {
+	b.callbacks.AddCallback(callback)
+}
+
+func (b *ReaderBatchWriter) Commit() error {
+	err := b.batch.Commit(pebble.Sync)
```

Comment:
when comparing the write performance between badger and pebble. I noticed pebble is 10x slower than badger. 
I tracked down to this `Commit` call by adding the following code. 

```
    n1 := time.Now()
    err := b.batch.Commit(pebble.Sync)
    n2 := time.Now()
    fmt.Println("pebbleimpl.Writer.go: Commit() time:", n2.Sub(n1).Nanoseconds())
```

If I do the same to badger to record the duration for the `Commit` call, and run the `TestReadWrite` test case which writes a single entity data to the database, the result shows pebble is 10x slower than badger for writes (read performance is similar). I'm not sure why. 

```
cd storage/operation
gt -run=TestReadWrite
badgerimpl.Writer.go: Commit() time: 147250
badgerimpl.Writer.go: Commit() time: 71458
badgerimpl.Writer.go: Commit() time: 72708
pebbleimpl.Writer.go: Commit() time: 3684875
pebbleimpl.Writer.go: Commit() time: 3006416
pebbleimpl.Writer.go: Commit() time: 3001083
badgerimpl.Writer.go: Commit() time: 6500
pebbleimpl.Writer.go: Commit() time: 584
PASS
ok      github.com/onflow/flow-go/storage/operation     1.136s
```

Reply:
Thanks for making the benchmarks and finding this regression @zhangchiqing 🙇 

100x slower writes is a big difference! 😬 

I think it would be worthwhile to spend some time better understanding this finding and investigating how it would impact Flow (for example, the consensus hot-path). My gut feeling is that this benchmark in isolation is overstating the actual impact we would see. If it isn't, and switching to Pebble is going to make all our writes 2 orders of magnitude slower, I'd rather know that now.
- Are there Pebble configs we can change to bring write speed down?
- Does this result hold if aspects of the environment are changed (eg. maybe it's an artifact of the test database being very small? Just guessing)
- Does this result hold if aspects of the load are changed (eg. committing a batch containing many writes)

I spent a short time during the review trying a few different iterations on `BenchmarkUpsert`:
- Setting `WithKeepL0InMemory` to false (previously was true). 
- Setting `WithValueThreshold` to 1 byte (previously was 1kb).
  - Badger is different from Pebble in that it will sometimes store values directly in the LSM tree, and other times in a value log file, based on the value's size. All the values we were storing were below that 1kb threshold. 
- Using larger 10kb values

Unfortunately, all of these had similar results where Pebble was still ~100x slower...

Reply:
Yeah, this might take some time to investigate. I will do it in a separate issue. But it should not block us from getting in with this PR, since it also allows us to move from badger transaction to badger batch updates. 

Regarding the regression, Peter suspect it might be to do with the key-compression and decompression, since he did a profiling in the past and saw a lot of CPU cycles are spent on that. So might worth to try disabling compression in pebble and see if it makes a difference.